src/main/scala/com.snowplowanalytics/iglu.schemaddl/migrations/SchemaList.scala

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright (c) 2014-2019 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+
+package com.snowplowanalytics.iglu.schemaddl.migrations
+
+// cats
+import cats.Monad
+import cats.data._
+import cats.implicits._
+import cats.kernel.Order
+
+import com.snowplowanalytics.iglu.core.{SchemaKey, SchemaList => SchemaKeyList, SchemaMap, SchemaVer}
+
+import com.snowplowanalytics.iglu.schemaddl.{IgluSchema, ModelGroup, VersionTree}
+import SchemaList.FullListCreationError._
+
+/**
+  * Always belong to the same vendor/name/model triple.
+  * Schemas in the list will be in correct order.
+  */
+sealed trait SchemaList extends Product with Serializable
+
+object SchemaList {
+
+  /**
+    * Always belong to the same vendor/name/model triple,
+    * always have at least two elements: head (initial x-0-0 schema) and latest schema
+    */
+  case class SchemaListFull private(schemas: NonEmptyList[IgluSchema]) extends SchemaList {
+
+    /**
+      * Create SchemaListSegment from schemas in current SchemaListFull
+      */
+    def toSegment: SchemaListSegment = SchemaListSegment(schemas)
+
+    /**
+      * Create segments from all possible combinations of NonEmptyList(source,between,destination),
+      */
+    def extractSegments: NonEmptyList[SchemaListSegment] = {
+      val res = SchemaList.buildMatrix(schemas.toList).map { case(_, _, nel) => SchemaListSegment(nel) }
+      NonEmptyList.fromListUnsafe(res)
+    }
+
+    /**
+      * Create new segment with items after index
+      */
+    def afterIndex(i: Int): Option[SchemaListSegment] =
```

Comment:
Can we make it `private[schemaddl]` or `private[migrations]`. I really dislike any APIs involving indexes, but don't see other solutions.